lib/ecto/query/builder/partition_by.ex

```diff
@@ -0,0 +1,49 @@
+import Kernel, except: [apply: 2]
+
+defmodule Ecto.Query.Builder.PartitionBy do
+  @moduledoc false
+
+  alias Ecto.Query.Builder
+  alias Ecto.Query.Builder.OrderBy
+
+  @doc """
+  """
+  @spec escape(Macro.t) :: {String.t, Macro.t}
+  def escape({:partition_by, _, [fields, opts]}) when is_list(opts), do: {escape_fields(fields), opts}
+  def escape({:partition_by, _, [fields]}), do: {escape_fields(fields), []}
+
+  def escape_fields(fields) when is_list(fields), do: fields
+  def escape_fields(field), do: [field]
+
+  def order_by(_, [], _), do: []
+  def order_by(binding, [{:order_by, expr} | rest], env) do
+    {expr, params} = OrderBy.escape(:order_by, expr, binding, env)
+    params = Builder.escape_params(params)
+
+    order_by = quote do: %Ecto.Query.QueryExpr{
+                       expr: unquote(expr),
+                       params: unquote(params),
+                       file: unquote(env.file),
+                       line: unquote(env.line)}
+
+    [order_by | order_by(binding, rest, env)]
+  end
+
+  @doc """
+  Builds a partition by expression.
+  """
+  @spec build([Macro.t], Tuple.t, Macro.Env.t) :: Macro.t
+  def build(binding, expr, env) do
+    {fields, opts} = escape(expr)
+    {fields, {_, :acc}} = Builder.escape(fields, :any, {%{}, :acc}, binding, env)
+    order_bys = order_by(binding, opts, env)
```

Comment:
I think this could be `Enum.map(opts, &order_by(&1, binding, env)` and we would be able to remove the loop from `order_by`?